File: TugasModul4/script/magnitude_server.py
#!/usr/bin/env python3

import rospy
from package1.srv import magnitude, magnitudeResponse
import math

def findMagnitude(req):
    # Mengolah & mencetak data
    resp = math.sqrt(req.komponenX**2 + req.komponenY**2 + req.komponenZ**2)
    rospy.loginfo("\n[Server]\nMENCARI MAGNITUDE\nKomponen X: %lf\nKomponen Y: %lf\nKomponen Z: %lf\nMagnitude: %f" 
                  % (req.komponenX, req.komponenY, req.komponenZ, resp))
    return magnitudeResponse(resp)

def magnitude_server():
    # Inisialisasi Node
    rospy.init_node("magnitude_server")
    
    # Insialisasi Server
    service = rospy.Service('magnitude_service', magnitude, findMagnitude)
    rospy.loginfo("Siap mencari magnitude!")
    rospy.spin()

if __name__ == '__main__':
    try:
        magnitude_server()
    except rospy.ROSInterruptException:
        pass

File: TugasModul4/src/package1/src/magnitude_server.cpp
#include "ros/ros.h"
#include "package1/magnitude.h"
#include <math.h>

bool findMagnitude(package1::magnitude::Request &req, package1::magnitude::Response &resp) {
    // Mengolah data request ke dalam respon
    resp.magnitude = sqrt(pow(req.komponenX, 2) + pow(req.komponenY, 2) + pow(req.komponenZ, 2));

    // Mencetak hasil olahan data
    ROS_INFO(
        "\n[Server]\n"
        "Nilai komponen X: %lf Nilai komponen Y: %lf Nilai komponen Z: %lf\n"
        "Magnitude: %lf\n",
        req.komponenX,
        req.komponenY,
        req.komponenZ,
        resp.magnitude
    );
    return true;
}

int main(int argc, char **argv)
{
    // Inisialisasi Node
    ros::init( argc, argv, "magnitude_server");

    // Inisialisasi Handler Node
    ros::NodeHandle nh;

    // Inisialisasi Server dari Handler
    ros::ServiceServer server = nh.advertiseService("magnitude", findMagnitude);

    ros::spin();

    return 0;
}
